people-and-planet-ai/timeseries-classification/e2e_test.py

```diff
@@ -0,0 +1,316 @@
+        ],
+        check=True,
+    )
+
+    # https://cloud.google.com/sdk/gcloud/reference/run/services/describe
+    service_url = (
+        subprocess.run(
+            [
+                "gcloud",
+                "run",
+                "services",
+                "describe",
+                service_name,
+                "--platform=managed",
+                f"--project={PROJECT}",
+                f"--region={REGION}",
+                "--format=get(status.url)",
+            ],
+            capture_output=True,
+        )
+        .stdout.decode("utf-8")
+        .strip()
+    )
+
+    logging.info(f"service_url: {service_url}")
+    yield service_url
+
+    # https://cloud.google.com/sdk/gcloud/reference/run/services/delete
+    subprocess.run(
+        [
+            "gcloud",
+            "run",
+            "services",
+            "delete",
+            service_name,
+            "--platform=managed",
+            f"--project={PROJECT}",
+            f"--region={REGION}",
+            "--quiet",
+        ],
+        check=True,
+    )
+
+
+@pytest.fixture(scope="session")
+def access_token() -> str:
+    yield (
+        subprocess.run(
+            ["gcloud", "auth", "print-identity-token", f"--project={PROJECT}"],
+            capture_output=True,
+        )
+        .stdout.decode("utf-8")
+        .strip()
+    )
+
+
+@pytest.fixture(scope="session")
+def create_datasets(
+    service_url: str, access_token: str, raw_data_dir: str, raw_labels_dir: str
+) -> str:
+
+    response = requests.post(
+        f"{service_url}/create-datasets",
+        headers={"Authorization": f"Bearer {access_token}"},
+        json={
+            "raw_data_dir": raw_data_dir,
+            "raw_labels_dir": raw_labels_dir,
+        },
+    ).json()
+    logging.info(f"create_datasets response: {response}")
+
+    job_id = response["job_id"]
+    job_url = response["job_url"]
+    logging.info(f"create_datasets job_id: {job_id}")
+    logging.info(f"create_datasets job_url: {job_url}")
+
+    # Wait until the Dataflow job finishes.
+    status = None
+    logging.info("Waiting for datasets to be created.")
+    for _ in range(0, TIMEOUT_SEC, POLL_INTERVAL_SEC):
+        # https://cloud.google.com/dataflow/docs/reference/rest/v1b3/projects.jobs/get
+        job = (
+            dataflow.projects()
+            .jobs()
+            .get(
+                projectId=PROJECT,
+                jobId=job_id,
+                view="JOB_VIEW_SUMMARY",
+            )
+            .execute()
+        )
+
+        status = job["currentState"]
+        if status in DATAFLOW_FINISHED_STATE:
+            break
+        time.sleep(POLL_INTERVAL_SEC)
+
+    logging.info(f"Datasets job finished with status {status}")
+    assert status == DATAFLOW_SUCCESS_STATE
+    yield job_id
+
+
+@pytest.fixture(scope="session")
+def train_model(service_url: str, access_token: str, create_datasets: str) -> str:
+    response = requests.post(
```

Comment:
Since you're making multiple requests with the same header I recommend `google.auth.transport.requests.AuthorizedSession`. This will refresh the credential pre-request as needed and attach the auth header.

I believe it will also let you remove the `access_token` fixture that calls out to gcloud.

https://googleapis.dev/python/google-auth/latest/reference/google.auth.transport.requests.html#google.auth.transport.requests.AuthorizedSession

```py
import google.auth
from google.auth.transport.requests import AuthorizedSession

credentials, _ = google.auth.default()
authed_session = AuthorizedSession(credentials)
response = authed_session.post(url="...", json="...")
```

Reply:
Nice! I didn't knew this existed! I'll give it a try! :)

Reply:
I got an error:

`'WWW-Authenticate': 'Bearer error="invalid_token" error_description="The access token could not be verified"'`

Cloud Run might be looking for something different in the request.

Reply:
I think the authenticated session might be using an "access token" while the Cloud Run service requires an "identity token". The notebook links to the
[Authentication overview](https://cloud.google.com/run/docs/authenticating/overview)
page for Cloud Run, which explains different methods of authenticating.

https://stackoverflow.com/questions/68655602/gcloud-identity-token-vs-access-token